search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,603 @@
+        return FALSE;
+
+    return g_dbus_interface_skeleton_export (G_DBUS_INTERFACE_SKELETON (self->skeleton),
+                                             connection,
+                                             object_path,
+                                             error);
+}
+
+static void
+yelp_search_provider_app_dbus_unregister (GApplication    *app,
+                                          GDBusConnection *connection,
+                                          const gchar     *object_path)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (app);
+    GDBusInterfaceSkeleton *skeleton = G_DBUS_INTERFACE_SKELETON (self->skeleton);
+
+    if (g_dbus_interface_skeleton_has_connection (skeleton, connection))
+        g_dbus_interface_skeleton_unexport_from_connection (skeleton, connection);
+
+    G_APPLICATION_CLASS (yelp_search_provider_app_parent_class)->dbus_unregister (app,
+                                                                                  connection,
+                                                                                  object_path);
+}
+
+static void
+yelp_search_provider_app_init (YelpSearchProviderApp *self)
+{
+    self->skeleton = yelp_shell_search_provider2_skeleton_new ();
+
+    g_signal_connect (self->skeleton, "handle-get-initial-result-set",
+                      G_CALLBACK (handle_get_initial_result_set), self);
+    g_signal_connect (self->skeleton, "handle-get-subsearch-result-set",
+                      G_CALLBACK (handle_get_subsearch_result_set), self);
+    g_signal_connect (self->skeleton, "handle-get-result-metas",
+                      G_CALLBACK (handle_get_result_metas), self);
+    g_signal_connect (self->skeleton, "handle-activate-result",
+                      G_CALLBACK (handle_activate_result), self);
+    g_signal_connect (self->skeleton, "handle-launch-search",
+                      G_CALLBACK (handle_launch_search), self);
+
+    g_application_set_inactivity_timeout (G_APPLICATION (self), 
+                                          SEARCH_PROVIDER_INACTIVITY_TIMEOUT);
+}
+
+static void
+yelp_search_provider_app_class_init (YelpSearchProviderAppClass *klass)
+{
+    GApplicationClass *aclass = G_APPLICATION_CLASS (klass);
+    GObjectClass *oclass = G_OBJECT_CLASS (klass);
+
+    aclass->startup = search_provider_app_startup;
+    aclass->dbus_register = yelp_search_provider_app_dbus_register;
+    aclass->dbus_unregister = yelp_search_provider_app_dbus_unregister;
+
+    oclass->dispose = search_provider_app_dispose;
+}
+
+static GApplication *
+yelp_search_provider_app_new (void)
+{
+    return g_object_new (YELP_TYPE_SEARCH_PROVIDER_APP,
+                         "application-id", "org.gnome.Yelp.SearchProvider",
+                         "flags", G_APPLICATION_IS_SERVICE,
+                         NULL);
+}
+
+int
+main (int    argc,
+      char **argv)
+{
+    GApplication *app;
+    gint res;
+    gchar *env_var = "YELP_SKIP_GUI";
```

Comment:
We were getting a crash, because libyelp is trying to get some info from GtkSettings and it needs a connection to display I think, which we obviously don't have. I don't remember the exact error we were getting here.

Probably the proper solution would actually to split libyelp into libyelp (uri handling, document handling, etc) and libyelp-gui (views, search bars, gui settings, etc), so the search provider could use the former only. But not sure how much work this would be.